### PR TITLE
[None][feat] Enable Marlin NVFP4 on Ada Lovelace

### PR DESCRIPTION
**Pull request by rmeghwal-nv**
## Description

Extend the existing opt-in Marlin NVFP4 path from Hopper-only coverage to also support Ada Lovelace / L40S (SM89), while preserving Hopper support.

This change compiles the Marlin NVFP4 kernels for SM89 and SM90, adds shared Ada/Hopper SM support gates in C++ and Python, and updates runtime checks for Marlin GEMM, fused MoE GEMM, and repack paths.

Marlin remains strictly opt-in. Users select it through `moe_config.backend: MARLIN` for MoE layers and `nvfp4_gemm_config.allowed_backends: ["marlin"]` for dense / Mamba Linear layers. Default Blackwell paths and non-Marlin fallback backends remain unchanged.

This PR also adds focused 1-GPU L40S coverage by reusing the existing Nemotron 3 Nano NVFP4 Marlin accuracy test with `tp_size=1`.

## Test Coverage

Added / updated coverage:

- Extended existing Marlin unit-test gates from Hopper-only to Ada/Hopper.
- Updated the existing Nemotron 3 Nano NVFP4 Marlin accuracy test so it can run on Ada/L40S with `tp_size=1`:
  - `accuracy/test_llm_api_pytorch.py::TestNemotronV3Nano::test_nvfp4_marlin_multi_gpus[tp_size=1]`
- Added the `tp_size=1` parametrization to the L40S pre-merge PyTorch lane:
  - `tests/integration/test_lists/test-db/l0_l40s.yml`

Validation run locally:

- Built `build_wheel_targets` in `tensorrt_llm/devel:devel-fresh-l40s`.
- Verified SM89 cubins with `cuobjdump` for:
  - `marlin_nvfp4_gemm.cu.o`
  - `marlin_nvfp4_moe_gemm.cu.o`
  - `marlin_repack.cu.o`
- Built wheel:
  - `build/tensorrt_llm-1.3.0rc21-cp312-cp312-linux_x86_64.whl`
- Verified imports inside the development container using the repo virtualenv:
  - TensorRT 10.16.1.11
  - PyTorch 2.12 NVIDIA container build
  - TensorRT-LLM 1.3.0rc21
- Passed targeted Marlin custom op test:
  - `tests/unittest/_torch/thop/serial/test_nvfp4_marlin_gemm.py`
  - Result: 48 passed, 2 warnings.
- Ran local Nemotron 3 Nano NVFP4 Marlin sanity checks:
  - Single GPU / TP=1 generation smoke passed with `max_batch_size=32`.
  - Single GPU / TP=1 generation smoke passed with `max_batch_size=128`.
  - Two GPU / TP=2 generation smoke passed.
  - Two GPU / TP=2 GSM8K sanity passed with score `87.5` on 8 samples.
- Validated test-list entries:
  - `python3 scripts/check_test_list.py --validate --test-lists-dir tests/integration/test_lists --test-base-dir tests/integration/defs`
  - Result: `OK: 2028 unique test entries validated.`
- Passed:
  - Python syntax compile for touched Python files.
  - `git diff --check`.

Notes:

- The L40S CI entry intentionally runs `tp_size=1` because the current `l0_l40s.yml` lane is `system_gpu_count: gte: 1, lte: 1`.
- Existing Hopper/H200 coverage continues to use `tp_size=8`.
- No new dependency is added.
- CODEOWNERS is unchanged.
- This PR focuses on architecture enablement and functional coverage for SM89, not on measured performance changes.

## PR Checklist

Please review the following before submitting your PR:

- [x] PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- [x] PR follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- [x] Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work)).
- [x] If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- [x] Any new dependencies have been scanned for license and vulnerabilities.
- [x] [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes.
- [ ] Documentation updated as needed.
- [x] Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- [x] The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- Extended opt-in Marlin NVFP4 support from Hopper SM90 to also cover Ada SM89 (L40S), while preserving Hopper behavior.
- Added/centralized SM support predicates in C++ and Python:
  - C++: compile-time device feature macro (`MARLIN_NVFP4_DEVICE_SUPPORTED`) and shared runtime helper (`marlin_nvfp4::isMarlinNvfp4SmSupported(int sm)`).
  - Python: shared helper `is_nvfp4_marlin_supported_sm()` and updated `is_nvfp4_marlin_enabled()` gating.
- Applied SM89–SM99 support consistently across Marlin NVFP4 paths:
  - Dense GEMM, fused MoE GEMM, repack, and linear methods (including updated error messages to reflect the expanded range).
- Updated CUDA build targets to compile Marlin NVFP4 kernels for both SM89 and SM90 (CMake `marlin_src` OBJECT library configuration).
- Kept Marlin behavior opt-in via the existing backend selection knobs (MoE `moe_config.backend: MARLIN` and NVFP4 gemm `allowed_backends: ["marlin"]`); broadened only the supported SM range under those explicit opt-in conditions.
- Documentation/update consistency:
  - Updated MoE developer guide and various module docstrings/comments to reflect Ada/Hopper NVFP4 W4A16 (SM89–SM99).

## QA Engineer Review

- Test-code changes (functions/classes modified):
  - `tests/integration/defs/accuracy/test_llm_api_pytorch.py`
    - `TestNemotronV3Nano.test_nvfp4_marlin_multi_gpus`: broadened skip gating to Ada, adjusted memory skip threshold, added MPI-world-size-aware skipping.
  - `tests/unittest/_torch/thop/parallel/test_fp4_linear.py`
    - `test_fp4_linear_marlin`: updated skip logic to use the shared Ada/Hopper SM support predicate.
  - `tests/unittest/_torch/thop/serial/test_nvfp4_marlin_gemm.py`
    - `TestNvfp4MarlinGemm` (decorators/skip gating): switched to Ada-or-Hopper-aware skip rule.
  - `tests/unittest/_torch/modules/moe/moe_test_utils.py`
    - Updated MARLIN backend skip logic to use the shared Ada/Hopper SM support predicate (no new test added; skip behavior changed).
- Test-list/config changes:
  - `tests/integration/test_lists/test-db/l0_l40s.yml`
    - Added accuracy test entry for `accuracy/test_llm_api_pytorch.py::TestNemotronV3Nano::test_nvfp4_marlin_multi_gpus[tp_size=1]` under the L40S single-GPU pre-merge PyTorch list.
- Coverage linkage:
  - The key Ada SM89 L40S coverage is represented in `test-db/l0_l40s.yml` via the added `TestNemotronV3Nano::test_nvfp4_marlin_multi_gpus[tp_size=1]` entry.
- **Verdict: sufficient**

### CODE REVIEW

**Review by StanleySun639 — APPROVED** [review]

**Comment by StanleySun639** [comment]
I noticed there is [perf] tag in the PR title but only added functionality test case. Do we need add perf coverage for it ?

**Review by BowenFu — APPROVED** [review]
Marlin NVFP4 enablement is cleanly gated: SM guards widen 90→{89,90} only inside Marlin-specific files, Marlin stays explicit opt-in (AUTO/default NVFP4 never selects it), and the device macro/CMake keep Hopper(SM90) and Blackwell(SM100+) unchanged. LGTM.

**Comment by rmeghwal-nv** [comment]
> I noticed there is [perf] tag in the PR title but only added functionality test case. Do we need add perf coverage for it ?

Good point. This PR is mainly architecture enablement with functional coverage, not a measured performance change. I updated the title from [perf] to [feat].

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py [hunk 1]
```diff
@@ -1153,12 +1153,12 @@ def forward(
...
         # Handle fallback tactic on cache miss
         if tactic == -1:
-            # Prefer marlin on Hopper (SM90) when explicitly allowed, cutlass
+            # Prefer marlin on Ada/Hopper (SM89-99) when explicitly allowed, cutlass
             # otherwise, falling back to whatever backend is available.
             assert len(
                 self.allowed_backends) > 0, "No allowed backends available"
             sm_version = get_sm_version()
-            if "marlin" in self.allowed_backends and 90 <= sm_version <= 99:
+            if "marlin" in self.allowed_backends and 89 <= sm_version <= 99:
```
Comment (xuantengh) [inline tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: I think we can utilize `from tensorrt_llm._torch.utils import is_nvfp4_marlin_supported_sm` and use this to check whether Marlin is supported based on the hardware compute capability in this file.

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py
Comment (xuantengh) [inline tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: ditto

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 2]
```diff
@@ -6895,6 +6895,27 @@ def test_nvfp4_marlin_multi_gpus(self, tp_size):
...
+                tensor_parallel_size=1,
+                moe_expert_parallel_size=1,
+                max_batch_size=8,
+                moe_config=MoeConfig(backend="MARLIN"),
+                nvfp4_gemm_config={"allowed_backends": ["marlin"]},
+        ) as llm:
+            task = MMLU(self.MODEL_NAME)
+            task.evaluate(llm,
+                          extra_evaluator_kwargs=self.EXTRA_EVALUATOR_KWARGS)
+
```
Comment (xuantengh) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py]: Do you think we can merge this with the `test_nvfp4_marlin_multi_gpus` above. We can add a parametrize indicating the number of GPUs, as the code looks similar in these 2 test functions.

File: tests/integration/test_lists/test-db/l0_l40s.yml [hunk 3]
```diff
@@ -36,6 +36,8 @@ l0_l40s:
   # AutoDeploy: Nemotron-Nano-V3 on Ada uses flashinfer; trtllm has no
   # (E4M3 input, BF16 output, paged_kv, head_dim=128, sm_89) FMHA cubin.
   - accuracy/test_llm_api_autodeploy.py::TestNemotronNanoV3::test_accuracy[fp8-1-attn_dp_off-flashinfer]
+  # PyTorch: force NVFP4 Marlin on Ada to cover the SM89-compiled Marlin kernels.
+  - accuracy/test_llm_api_pytorch.py::TestNemotronV3Nano::test_nvfp4_marlin_single_gpu
```
Comment (xuantengh) [inline tests/integration/test_lists/test-db/l0_l40s.yml]: Update the test name here if you decide to merge the test functions.

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py [hunk 4]
```diff
@@ -751,8 +751,8 @@ def test_fp4_linear_cuda_core(dtype, mnk):
 
 
 @pytest.mark.skipif(
-    get_sm_version() < 90 or get_sm_version() >= 100,
-    reason="Marlin NVFP4 backend runs Hopper",
+    not (89 <= get_sm_version() < 100),
```
Comment (xuantengh) [inline tests/unittest/_torch/thop/parallel/test_fp4_linear.py]: Similarly, we can use `is_nvfp4_marlin_supported_sm` to check the Marlin and SM match so no hard-coded SM number everywhere.

File: cpp/tensorrt_llm/kernels/marlin/marlin_nvfp4.h [hunk 5]
```diff
@@ -25,6 +25,11 @@
 namespace marlin_nvfp4
 {
 
+inline bool isMarlinNvfp4SmSupported(int sm)
+{
+    return sm >= 89 && sm < 100;
+}
```
[the remark this answers is not in the document]
Comment (rmeghwal-nv) [reply cpp/tensorrt_llm/kernels/marlin/marlin_nvfp4.h:31]: Verified against the current branch. There is only one definition of isMarlinNvfp4SmSupported, in marlin_nvfp4.h. The three .cu files only call the helper and do not define it, so no code change is needed here.

**Review by xuantengh — CHANGES_REQUESTED** [review]

File: tensorrt_llm/_torch/modules/fused_moe/create_moe.py [hunk 6]
```diff
@@ -64,7 +64,7 @@ def get_moe_cls(
         quant_config = override_quant_config
     layer_prefix = f"[layer_idx={layer_idx}] " if layer_idx is not None else ""
     if moe_backend.upper() == "MARLIN":
-        # Marlin MoE is a Hopper-specific NVFP4 W4A16 backend. Require nvfp4
```
Comment (xxi-nv) [inline tensorrt_llm/_torch/modules/fused_moe/create_moe.py]: Please help to update the MOE_DEVELOPER_GUIDE.md.

**Review by xxi-nv — APPROVED** [review]

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py (hunk 1 above)
Comment (rmeghwal-nv) [reply tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: fixed.

File: tensorrt_llm/_torch/custom_ops/torch_custom_ops.py
Comment (rmeghwal-nv) [reply tensorrt_llm/_torch/custom_ops/torch_custom_ops.py]: fixed.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 2 above)
Comment (rmeghwal-nv) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: I have removed this test, and updated old test only ( changed skip code)

File: tests/integration/test_lists/test-db/l0_l40s.yml (hunk 3 above)
Comment (rmeghwal-nv) [reply tests/integration/test_lists/test-db/l0_l40s.yml]: updated.

File: tests/unittest/_torch/thop/parallel/test_fp4_linear.py (hunk 4 above)
Comment (rmeghwal-nv) [reply tests/unittest/_torch/thop/parallel/test_fp4_linear.py]: updated.

**Comment by xuantengh** [comment]
@rmeghwal-nv Please re-run the `pre-commit` hooks to format the changes.

**Review by litaotju — APPROVED** [review]

**Review by xuantengh — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fused_moe/create_moe.py (hunk 6 above)
Comment (rmeghwal-nv) [reply tensorrt_llm/_torch/modules/fused_moe/create_moe.py]: updated.

File: tensorrt_llm/_torch/modules/fused_moe/MOE_DEVELOPER_GUIDE.md [hunk 7]
```diff
@@ -150,7 +150,7 @@ Still on old path (standalone, with embedded communication):
...
-| `fused_moe_marlin.py` | `MarlinFusedMoE` | SM90 only | W4A16 NVFP4 on Hopper (BF16 activations + FP4 weights, fused single-launch `marlin_nvfp4_moe_gemm` kernel); supports attention-DP + EP via external comm (scheduler precomputes routing; dispatch payload is plain BF16, no activation scales); non-NVFP4 layers (e.g. unquantized MTP draft layers) fall back to Cutlass in `get_moe_cls`; no dynamic EPLB | `EXTERNAL_COMM` |
+| `fused_moe_marlin.py` | `MarlinFusedMoE` | SM89-SM99 | W4A16 NVFP4 on Ada/Hopper (BF16 activations + FP4 weights, fused single-launch `marlin_nvfp4_moe_gemm` kernel); supports attention-DP + EP via external comm (scheduler precomputes routing; dispatch payload is plain BF16, no activation scales); non-NVFP4 layers (e.g. unquantized MTP draft layers) fall back to Cutlass in `get_moe_cls`; no dynamic EPLB | `EXTERNAL_COMM` |
```
[the remark this answers is not in the document]
Comment (rmeghwal-nv) [reply tensorrt_llm/_torch/modules/fused_moe/MOE_DEVELOPER_GUIDE.md:153]: updated

File: tests/integration/test_lists/test-db/l0_l40s.yml [hunk 8]
```diff
@@ -36,6 +36,8 @@ l0_l40s:
   # AutoDeploy: Nemotron-Nano-V3 on Ada uses flashinfer; trtllm has no
   # (E4M3 input, BF16 output, paged_kv, head_dim=128, sm_89) FMHA cubin.
   - accuracy/test_llm_api_autodeploy.py::TestNemotronNanoV3::test_accuracy[fp8-1-attn_dp_off-flashinfer]
+  # PyTorch: force NVFP4 Marlin on Ada to cover the SM89-compiled Marlin kernels.
+  - accuracy/test_llm_api_pytorch.py::TestNemotronV3Nano::test_nvfp4_marlin_multi_gpus[tp_size=1]
```
Comment (xinhe-nv) [inline tests/integration/test_lists/test-db/l0_l40s.yml:40]: could you add this case into QA test list, too?

Comment (rmeghwal-nv) [reply tests/integration/test_lists/test-db/l0_l40s.yml:40]: added test to tests/integration/test_lists/qa/llm_function_core.txt
